### PR TITLE
Adding support to Custom Aggregation Methods

### PR DESCRIPTION
**Pull request by athoscouto**
### Description
*Creating CustomAggregateMethodAnnotation class that allows client to notify OData about implemented custom aggregate methods. Modifying AggregationBinder to support Custom Aggregation Methods according to OData 4.0 spec. Right now only methods that return doubles are supported*

### Checklist (Uncheck if it is not completed)
- [x] Test cases added
- [ ] Build and test with one-click build and test script passed

### Additional work necessary
*Basics tests are implemented, maybe it would be nice to have more thorough tests.*

### CODE REVIEW

File: OData/src/System.Web.OData/OData/CustomAggregateMethodAnnotation.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        {
+            TokenToMethodWrapper.Add(methodToken, methods);
+            return this;
+        }
+
+        /// <summary>
+        /// Get an implementation of a method with the specifieds returnType and methodToken.
+        /// If there's no method that matches the requirements, returns null.
+        /// </summary>
+        public bool getMethodInfo(string methodToken, Type returnType, out MethodInfo method)
```
Comment (kosinsky) [inline OData/src/System.Web.OData/OData/CustomAggregateMethodAnnotation.cs]: In C# Pascal casing is used for public methods.

File: OData/src/System.Web.OData/OData/CustomAggregateMethodAnnotation.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    /// <summary>
+    /// Allows client to tell OData which are the custom aggregation methods defined.
+    /// In order to do it it must receive a methodToken - that is the full identifier
+    /// of the method in the OData URL - and an IDictionary that maps the input type
+    /// of the aggregation method to it's MethodInfo.
+    /// </summary>
+    public class CustomAggregateMethodAnnotation 
+    {
+
+        Dictionary<string, IDictionary<Type, MethodInfo>> TokenToMethodWrapper;
```
Comment (kosinsky) [inline OData/src/System.Web.OData/OData/CustomAggregateMethodAnnotation.cs]: private and _tokenToMethodWrapper

File: OData/src/System.Web.OData/OData/CustomAggregateMethodAnnotation.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+        /// <summary>
+        /// CustomAggregateMethodAnnotation simple constructor.
+        /// </summary>
+        public CustomAggregateMethodAnnotation()
+        {
+            TokenToMethodWrapper = new Dictionary<string, IDictionary<Type, MethodInfo>>();
```
Comment (kosinsky) [inline OData/src/System.Web.OData/OData/CustomAggregateMethodAnnotation.cs]: You could to that initialization in the same line as definition of the property.

File: OData/src/System.Web.OData/OData/Query/Expressions/AggregationBinder.cs [hunk 4]
```diff
@@ -181,72 +181,94 @@ private Expression CreateAggregationExpression(ParameterExpression accum, Aggreg
...
+            switch (expression.Method.MethodType)
             {
-                case AggregationMethod.Min:
-                {
-                    var minMethod = ExpressionHelperMethods.QueryableMin.MakeGenericMethod(this._elementType,
-                        propertyLambda.Body.Type);
-                    aggregationExpression = Expression.Call(null, minMethod, asQuerableExpression, propertyLambda);
-                }
+                case AggregationMethodType.Min:
+                    {
```
Comment (kosinsky) [inline OData/src/System.Web.OData/OData/Query/Expressions/AggregationBinder.cs]: Looks like just extra spaces

File: OData/src/System.Web.OData/OData/CustomAggregateMethodAnnotation.cs (hunk 2 above)
Comment (athoscouto) [reply OData/src/System.Web.OData/OData/CustomAggregateMethodAnnotation.cs]: Changed to private and _tokenToMethodMap.

File: OData/src/System.Web.OData/OData/CustomAggregateMethodAnnotation.cs (hunk 3 above)
Comment (athoscouto) [reply OData/src/System.Web.OData/OData/CustomAggregateMethodAnnotation.cs]: Done.

File: OData/src/System.Web.OData/OData/CustomAggregateMethodAnnotation.cs (hunk 1 above)
Comment (athoscouto) [reply OData/src/System.Web.OData/OData/CustomAggregateMethodAnnotation.cs]: Done, thanks for the tips!

File: OData/src/System.Web.OData/OData/Query/Expressions/AggregationBinder.cs (hunk 4 above)
Comment (athoscouto) [reply OData/src/System.Web.OData/OData/Query/Expressions/AggregationBinder.cs]: Removed them.

**Pull request closed without merging**